### PR TITLE
Make IPropertySubpattern Internal

### PR DESCRIPTION
**Pull request by 333fred**
Updates the structure of IPropertySubpattern to be what we decided in the IOperation design meeting, and makes the implementation internal until we finalize C# 8.
Fixes https://github.com/dotnet/roslyn/issues/33018.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs [hunk 1]
```diff
@@ -57,7 +58,7 @@ public static ValueUsageInfo GetValueUsageInfo(this IOperation operation)
                         //
                         return ValueUsageInfo.Write;
 
-                    case IRecursivePatternOperation _:
+                    case IOperation iop when iop.GetType().GetInterfaces().Any(i => i.Name == "IRecursivePatternOperation"):
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs]: Thanks for updating this. Can you also file a tracking issue to revert this IDE change?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 2]
```diff
@@ -2007,5 +2008,42 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
...
+                case PropertySymbol property:
+                    {
+                        var receiver = new InstanceReferenceOperation(
+                            InstanceReferenceKind.PatternInput, _semanticModel, nameSyntax, property.ContainingType, constantValue: default, isImplicit: true);
+                        return new PropertyReferenceOperation(
+                            property, receiver, ImmutableArray<IArgumentOperation>.Empty, _semanticModel, nameSyntax, property.Type.TypeSymbol,
+                            constantValue: default, isImplicit: isImplicit);
+                    }
+                case null:
+                    return null;
```
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Should this also return an invalid operation? At least the doc comment [here](https://github.com/dotnet/roslyn/pull/33178/files#diff-ecc8c3941cade7f6eb4c8c01f88a46c1R16) indicates Member is always non-null. If we do decide to allow null, we should make this explicit in doc comments.

File: src/Compilers/Core/Portable/Operations/OperationCloner.cs [hunk 3]
```diff
@@ -549,6 +549,16 @@ public override IOperation VisitRecursivePattern(IRecursivePatternOperation oper
                 operation.IsImplicit);
         }
 
+        internal override IOperation VisitPropertySubpattern(IPropertySubpatternOperation operation, object argument)
+        {
+            return new PropertySubpatternOperation(
+                semanticModel: null,
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/OperationCloner.cs]: >semanticModel: null, [](start = 16, length = 20)

Semantic model should be preserved by the Cloner.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 4]
```diff
@@ -6547,20 +6549,30 @@ public override IOperation VisitDeclarationPattern(IDeclarationPatternOperation
...
         }
 
+        internal override IOperation VisitPropertySubpattern(IPropertySubpatternOperation operation, int? argument)
+        {
+            return new PropertySubpatternOperation(
+                semanticModel: null,
+                operation.Syntax,
+                IsImplicit(operation),
+                Visit(operation.Member),
+                (IPatternOperation)Visit(operation.Pattern));
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >(IPatternOperation)Visit(operation.Pattern) [](start = 16, length = 43)

It looks like in this case we are not setting the right substitution for the ```InstanceReferenceKind.PatternInput``` for the case when the ```operation.Pattern``` is a recursive pattern. Given that majority of nodes are internal, I'd suggest to not substitute ```InstanceReferenceKind.PatternInput``` with a capture in this PR and leave it as a regular node. We will have to figure out the right way to deal with this later.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs [hunk 5]
```diff
@@ -739,17 +766,19 @@ void M((int X, int Y)? x, bool b)
                                   ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1) (Syntax: '1')
                               IDiscardPatternOperation (OperationKind.DiscardPattern, Type: null) (Syntax: '_') (InputType: System.Int32)
                           PropertySubpatterns (1):
-                            MatchedSymbol: System.Int32 (System.Int32 X, System.Int32 Y).Item1, Pattern: 
-                                IDeclarationPatternOperation (OperationKind.DeclarationPattern, Type: null) (Syntax: 'var z') (InputType: System.Int32, DeclaredSymbol: System.Int32 z, MatchesNull: True)
-
+                              IPropertySubpatternOperation (OperationKind.None, Type: null) (Syntax: 'Item1: var z')
+                                Member: 
+                                  IFieldReferenceOperation: System.Int32 (System.Int32 X, System.Int32 Y).Item1 (OperationKind.FieldReference, Type: System.Int32, IsImplicit) (Syntax: 'Item1')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs]: >IsImplicit [](start = 163, length = 10)

Is this supposed to be implicit?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 6]
```diff
@@ -2007,5 +2008,42 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
...
+        internal IPropertySubpatternOperation CreatePropertySubpattern(BoundSubpattern subpattern)
+        {
+            SyntaxNode syntax = subpattern.Syntax;
+            return new CSharpLazyPropertySubpatternOperation(this, subpattern, syntax, _semanticModel);
+        }
+
+        internal IOperation CreatePropertySubpatternMember(Symbol symbol, SyntaxNode syntax)
+        {
+            var nameSyntax = (syntax is SubpatternSyntax subpatSyntax ? subpatSyntax.NameColon?.Name : null) ?? syntax;
+            bool isImplicit = nameSyntax != syntax;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >nameSyntax != syntax [](start = 30, length = 20)

Should this condition be inversed? It looks like we are lacking a test that simply asserts IOperation tree shape for this code path, consider adding one.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 7]
```diff
@@ -2007,5 +2008,42 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
...
+            switch (symbol)
+            {
+                case FieldSymbol field:
+                    {
+                        var constantValue = field.ConstantValue is null ? default(Optional<object>) : new Optional<object>(field.ConstantValue);
+                        var receiver = new InstanceReferenceOperation(
+                            InstanceReferenceKind.PatternInput, _semanticModel, nameSyntax, field.ContainingType, constantValue, isImplicit: true);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >field.ContainingType [](start = 92, length = 20)

It feels like '''field.ContainingType``` is not always the right value to use. The field can come from a base type. I think we would want all PatternInputs within the same pattern to use the same type, matching the ```MatchedType```.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 8]
```diff
@@ -2007,5 +2008,42 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
...
+                        return new FieldReferenceOperation(
+                            field, isDeclaration: false, receiver, _semanticModel, nameSyntax, field.Type.TypeSymbol, constantValue, isImplicit: isImplicit);
+                    }
+                case PropertySymbol property:
+                    {
+                        var receiver = new InstanceReferenceOperation(
+                            InstanceReferenceKind.PatternInput, _semanticModel, nameSyntax, property.ContainingType, constantValue: default, isImplicit: true);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >property.ContainingType [](start = 92, length = 23)

Same comment.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 9]
```diff
@@ -2007,5 +2008,42 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
...
+                        return new PropertyReferenceOperation(
+                            property, receiver, ImmutableArray<IArgumentOperation>.Empty, _semanticModel, nameSyntax, property.Type.TypeSymbol,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >ImmutableArray<IArgumentOperation>.Empty [](start = 48, length = 40)

In an error case, could we ran into a parameterized property?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs [hunk 10]
```diff
@@ -1501,24 +1504,50 @@ public CSharpLazyRecursivePatternOperation(CSharpOperationFactory operationFacto
...
+    {
+        private readonly BoundSubpattern _subpattern;
+        private readonly CSharpOperationFactory _operationFactory;
+
+        public CSharpLazyPropertySubpatternOperation(
+            CSharpOperationFactory operationFactory,
+            BoundSubpattern subpattern,
+            SyntaxNode syntax,
+            SemanticModel semanticModel)
+            : base(semanticModel, syntax, false)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs]: >false [](start = 42, length = 5)

Is this for Implicit? Consider using named parameter.

File: src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs [hunk 11]
```diff
@@ -1186,24 +1186,34 @@ public override void VisitRecursivePattern(IRecursivePatternOperation operation)
...
             IEnumerable<IOperation> children = operation.DeconstructionSubpatterns.Cast<IOperation>();
-            children = children.Concat(operation.PropertySubpatterns.Select(x => x.Item2));
+            children = children.Concat(operation.PropertySubpatterns);
+
+            AssertEx.Equal(children, operation.Children);
+        }
+
+        internal override void VisitPropertySubpattern(IPropertySubpatternOperation operation)
+        {
+            Assert.True(operation.Member is IMemberReferenceOperation);
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: >Assert.True(operation.Member is IMemberReferenceOperation); [](start = 12, length = 59)

What about error cases?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 9 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: It does not appear so. The closest I could get was an `IConstantPatternOperation` child of the `IPropertyPatternOperation`. Using an indexer like below doesn't parse close to cleanly at all, it results in multiple `IncompleteMember`s on the class.
```C#
class C
{
    (int x, int y) this[int i] { get => default; set {} }
    void M(object o, bool b)
    {
        b = /*<bind>*/o is C { [1]: (var x, var y) }/*</bind>*/;
    }
}
```

---
In reply to: [254445327](https://github.com/dotnet/roslyn/pull/33178#discussion_r254445327) [](ancestors = 254445327)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs (hunk 4 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Added a note to [https://github.com/dotnet/roslyn/issues/33175](https://github.com/dotnet/roslyn/issues/33175) to follow up.

---
In reply to: [254440669](https://github.com/dotnet/roslyn/pull/33178#discussion_r254440669) [](ancestors = 254440669)

File: src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs (hunk 1 above)
Comment (333fred) [reply src/Workspaces/Core/Portable/Extensions/OperationExtensions.cs]: Added a note to [https://github.com/dotnet/roslyn/issues/33175](https://github.com/dotnet/roslyn/issues/33175).

---
In reply to: [254423185](https://github.com/dotnet/roslyn/pull/33178#discussion_r254423185) [](ancestors = 254423185)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [461185324](https://github.com/dotnet/roslyn/pull/33178#issuecomment-461185324) [](ancestors = 461185324)

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs [hunk 12]
```diff
@@ -1501,24 +1504,53 @@ public CSharpLazyRecursivePatternOperation(CSharpOperationFactory operationFacto
...
         public override ImmutableArray<IPatternOperation> CreateDeconstructionSubpatterns()
         {
             return _boundRecursivePattern.Deconstruction.IsDefault ? ImmutableArray<IPatternOperation>.Empty :
                 _boundRecursivePattern.Deconstruction.SelectAsArray((p, fac) => (IPatternOperation)fac.Create(p.Pattern), _operationFactory);
         }
-        public override ImmutableArray<(ISymbol, IPatternOperation)> CreatePropertySubpatterns()
+        public override ImmutableArray<IPropertySubpatternOperation> CreatePropertySubpatterns()
+        {
+            return _boundRecursivePattern.Properties.IsDefault ? ImmutableArray<IPropertySubpatternOperation>.Empty :
+                _boundRecursivePattern.Properties.SelectAsArray((p, fac) => fac.CreatePropertySubpattern(p, MatchedType), _operationFactory);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationNodes.cs]: >MatchedType [](start = 108, length = 11)

Is this going to prevent the lambda from caching the delegate? Consider passing it in a tuple parameter, together with operation factory.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 13]
```diff
@@ -2007,5 +2008,40 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
...
+                        var receiver = new InstanceReferenceOperation(
+                            InstanceReferenceKind.PatternInput, _semanticModel, nameSyntax, matchedType, constantValue: default, isImplicit: true);
+                        return new PropertyReferenceOperation(
+                            property, receiver, ImmutableArray<IArgumentOperation>.Empty, _semanticModel, nameSyntax, property.Type.TypeSymbol,
+                            constantValue: default, isImplicit: isImplicit);
+                    }
+                default:
+                    // We should expose the symbol in this case somehow:
+                    // https://github.com/dotnet/roslyn/issues/33175
+                    return OperationFactory.CreateInvalidOperation(_semanticModel, nameSyntax, ImmutableArray<IOperation>.Empty, isImplicit: true);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >true [](start = 141, length = 4)

Should we use ```isImplicit``` instead of hard coding ```true```?

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs [hunk 14]
```diff
@@ -882,14 +1002,135 @@ void M((int X, int Y) tuple, bool b)
...
+          IConstantPatternOperation (OperationKind.ConstantPattern, Type: null) (Syntax: '1') (InputType: System.Int32)
+            Value: 
+              ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1) (Syntax: '1')
+          IConstantPatternOperation (OperationKind.ConstantPattern, Type: null) (Syntax: '2') (InputType: System.Int32)
+            Value: 
+              ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 2) (Syntax: '2')
+      PropertySubpatterns (1):
+          IPropertySubpatternOperation (OperationKind.None, Type: null, IsInvalid) (Syntax: 'NotFound: int x')
+            Member: 
+              IInvalidOperation (OperationKind.Invalid, Type: null, IsInvalid, IsImplicit) (Syntax: 'NotFound')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs]: >IsImplicit [](start = 79, length = 10)

It feels like this shouldn't be implicit.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs [hunk 15]
```diff
@@ -882,14 +1002,135 @@ void M((int X, int Y) tuple, bool b)
...
+";
+
+            var vbCompilation = CreateVisualBasicCompilation(vbSource);
+
+            var source = @"
+class C
+{
+    void M1(object o, bool b)
+    {
+        b = /*<bind>*/o is C1 { Prop[1]: var x }/*</bind>*/;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs]: >Prop[1] [](start = 32, length = 7)

Consider adding a test without indexing, just the property name

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs [hunk 16]
```diff
@@ -882,14 +1002,135 @@ void M((int X, int Y) tuple, bool b)
...
+      DeconstructionSubpatterns (0)
+      PropertySubpatterns (2):
+          IPropertySubpatternOperation (OperationKind.None, Type: null, IsInvalid) (Syntax: 'Prop[1]')
+            Member: 
+              IInvalidOperation (OperationKind.Invalid, Type: null, IsInvalid, IsImplicit) (Syntax: 'Prop[1]')
+                Children(0)
+            Pattern: 
+              IConstantPatternOperation (OperationKind.ConstantPattern, Type: null, IsInvalid) (Syntax: 'Prop[1]') (InputType: ?)
+                Value: 
+                  IInvalidOperation (OperationKind.Invalid, Type: ?, IsInvalid) (Syntax: 'Prop[1]')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs]: >'Prop[1]' [](start = 89, length = 9)

Is this the same syntax node as for the IInvalidOperation above?

File: src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs [hunk 17]
```diff
@@ -1186,27 +1186,35 @@ public override void VisitRecursivePattern(IRecursivePatternOperation operation)
...
-            children = children.Concat(operation.PropertySubpatterns.Select(x => x.Item2));
+            children = children.Concat(operation.PropertySubpatterns);
 
             AssertEx.Equal(children, operation.Children);
         }
 
+        internal override void VisitPropertySubpattern(IPropertySubpatternOperation operation)
+        {
+            Assert.True(operation.Member is IMemberReferenceOperation || operation.Member is IInvalidOperation);
+            var member = (IMemberReferenceOperation)operation.Member;
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: >(IMemberReferenceOperation)operation.Member; [](start = 25, length = 44)

Is this going to throw when ```operation.Member is IInvalidOperation```?

Comment (AlekseyTs) [reply src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: Is this not getting caught by tests? Even when IOperation hook is enabled?

---
In reply to: [254480084](https://github.com/dotnet/roslyn/pull/33178#discussion_r254480084) [](ancestors = 254480084)

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs (hunk 16 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs]: The syntax of the top `IPropertySubpatternOperation` and the `IInvalidOperation` that is it's `Member` are the same, they are `SubpatternSyntax` nodes. These lower 2, the `IConstantPatternOperation` and the `IInvalidOperation` you commented on, are `ConstantPatternSyntax` and `ElementAccessExpressionSyntax` nodes, respectively.

---
In reply to: [254478564](https://github.com/dotnet/roslyn/pull/33178#discussion_r254478564) [](ancestors = 254478564)

**Comment by AlekseyTs** [comment]
It looks like there are legitimate test failures.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3)

File: src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs (hunk 17 above)
Comment (333fred) [reply src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: I should probably run these tests through the hook, I haven't. But no, this visitor isn't called on our regular IOperation tests.

---
In reply to: [254480510](https://github.com/dotnet/roslyn/pull/33178#discussion_r254480510) [](ancestors = 254480510,254480084)

**Comment by 333fred** [comment]
There were on the first run, second run hasn't completed yet.

---
In reply to: [461217005](https://github.com/dotnet/roslyn/pull/33178#issuecomment-461217005) [](ancestors = 461217005)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [461217061](https://github.com/dotnet/roslyn/pull/33178#issuecomment-461217061) [](ancestors = 461217061)

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs [hunk 18]
```diff
@@ -882,14 +1002,186 @@ void M((int X, int Y) tuple, bool b)
...
+{
+    void M1(object o, bool b)
+    {
+        b = /*<bind>*/o is C1 { Prop: var x }/*</bind>*/;
+    }
+}";
+
+            var compilation = CreateCompilation(source, new[] { vbCompilation.EmitToImageReference() });
+            compilation.VerifyDiagnostics(
+                // (6,33): error CS0154: The property or indexer 'Prop' cannot be used in this context because it lacks the get accessor
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs]: >// (6,33): error CS0154: The property or indexer 'Prop' cannot be used in this context because it lacks the get accessor [](start = 16, length = 120)

This error feels misleading, there is a get accessor. Consider opening an issue to follow up.

File: src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs [hunk 19]
```diff
@@ -1186,25 +1186,38 @@ public override void VisitRecursivePattern(IRecursivePatternOperation operation)
...
+                return;
+            }
+
+            Assert.True(operation.Member is IMemberReferenceOperation);
+            var member = (IMemberReferenceOperation)operation.Member;
+            switch (member.Member)
+            {
+                case IFieldSymbol field:
+                case IPropertySymbol prop:
+                case IErrorTypeSymbol error:
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: >case IErrorTypeSymbol error: [](start = 16, length = 28)

This is probably not a valid case any more.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs (hunk 18 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs]: Filled https://github.com/dotnet/roslyn/issues/33189.

File: src/Compilers/Core/Portable/Operations/OperationNodes.cs [hunk 20]
```diff
@@ -8089,6 +8088,100 @@ public override ImmutableArray<IPatternOperation> DeconstructionSubpatterns
...
+    }
+
+    internal sealed partial class PropertySubpatternOperation : BasePropertySubpatternOperation
+    {
+        public PropertySubpatternOperation(
+            SemanticModel semanticModel,
+            SyntaxNode syntax,
+            bool isImplicit,
+            IOperation member,
+            IPatternOperation pattern)
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/OperationNodes.cs]: Minor point: It looks like constructors for other `IOperation` implementations have child arguments before `semanticModel, syntax, isImplicit` arguments.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 21]
```diff
@@ -2007,5 +2008,40 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
...
+
+        internal IPropertySubpatternOperation CreatePropertySubpattern(BoundSubpattern subpattern, ITypeSymbol matchedType)
+        {
+            SyntaxNode syntax = subpattern.Syntax;
+            return new CSharpLazyPropertySubpatternOperation(this, subpattern, matchedType, syntax, _semanticModel);
+        }
+
+        internal IOperation CreatePropertySubpatternMember(Symbol symbol, ITypeSymbol matchedType, SyntaxNode syntax)
+        {
+            var nameSyntax = (syntax is SubpatternSyntax subpatSyntax ? subpatSyntax.NameColon?.Name : null) ?? syntax;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >(syntax is SubpatternSyntax subpatSyntax ? subpatSyntax.NameColon?.Name : null) [](start = 29, length = 79)

Could remove the `: null` with `(syntax as SubpatternSyntax)?.NameColon?.Name

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 22]
```diff
@@ -2007,5 +2008,40 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
...
+            bool isImplicit = nameSyntax == syntax;
+            switch (symbol)
+            {
+                case FieldSymbol field:
+                    {
+                        var constantValue = field.ConstantValue is null ? default(Optional<object>) : new Optional<object>(field.ConstantValue);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >field.ConstantValue [](start = 123, length = 19)

Should this be `field.ContantValue.Value`?

If so, please use the helper method:
```
var constantValue = ConvertToOptional(field.ConstantValue);
```

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 23]
```diff
@@ -2007,5 +2008,40 @@ private IOperation CreateUsingLocalDeclarationsOperation(BoundUsingLocalDeclarat
...
+                        var receiver = new InstanceReferenceOperation(
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >InstanceReferenceOperation [](start = 43, length = 26)

Are we testing with static fields and properties?
```
class C
{
    static int F;

    static void M(object o)
    {
        bool b = o is C { F: 1 };
    }
}
```

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs [hunk 24]
```diff
@@ -14,7 +14,7 @@ internal sealed partial class CSharpOperationFactory
     {
         private static readonly IConvertibleConversion s_boxedIdentityConversion = Conversion.Identity;
 
-        private static Optional<object> ConvertToOptional(ConstantValue value)
+        internal static Optional<object> ConvertToOptional(ConstantValue value)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs]: >internal [](start = 8, length = 8)

Is this change necessary?

**Comment by 333fred** [comment]
@cston none of these comments seem blocking, so I've added notes to https://github.com/dotnet/roslyn/issues/33175 to follow up.
